llvm/lib/Transforms/Utils/SYCLModuleSplit.cpp

```diff
@@ -0,0 +1,935 @@
+// functions and globals that can be called from entry points that are specified
+// in ModuleEntryPoints vector, in addition to the entry point functions.
+ModuleDesc extractCallGraph(const ModuleDesc &MD,
+                            EntryPointGroup ModuleEntryPoints,
+                            const DependencyGraph &CG,
+                            const std::function<bool(const Function *)>
+                                &IncludeFunctionPredicate = nullptr) {
+  SetVector<const GlobalValue *> GVs;
+  collectFunctionsAndGlobalVariablesToExtract(
+      GVs, MD.getModule(), ModuleEntryPoints, CG, IncludeFunctionPredicate);
+
+  ModuleDesc SplitM = extractSubModule(MD, GVs, std::move(ModuleEntryPoints));
+  LLVM_DEBUG(SplitM.dump(dbgs()));
+  SplitM.cleanup();
+
+  return SplitM;
+}
+
+class ModuleCopier : public ModuleSplitterBase {
+public:
+  using ModuleSplitterBase::ModuleSplitterBase; // to inherit base constructors
+
+  ModuleDesc nextSplit() override {
+    ModuleDesc Desc{releaseInputModule(), nextGroup()};
+    // Do some basic optimization like unused symbol removal
+    // even if there was no split.
+    Desc.cleanup();
+    return Desc;
+  }
+};
+
+class ModuleSplitter : public ModuleSplitterBase {
+public:
+  ModuleSplitter(ModuleDesc MD, EntryPointGroupVec GroupVec)
+      : ModuleSplitterBase(std::move(MD), std::move(GroupVec)),
+        CG(Input.getModule()) {}
+
+  ModuleDesc nextSplit() override {
+    return extractCallGraph(Input, nextGroup(), CG);
+  }
+
+private:
+  DependencyGraph CG;
+};
+
+} // namespace
+
+namespace llvm {
+
+std::optional<IRSplitMode> convertStringToSplitMode(StringRef S) {
+  static const StringMap<IRSplitMode> Values = {{"kernel", SPLIT_PER_KERNEL},
+                                                {"source", SPLIT_PER_TU},
+                                                {"auto", SPLIT_AUTO},
+                                                {"none", SPLIT_NONE}};
+
+  auto It = Values.find(S);
+  if (It == Values.end())
+    return std::nullopt;
+
+  return It->second;
+}
+
+void dumpEntryPoints(raw_ostream &OS, const EntryPointSet &C,
+                     std::string_view Msg) {
+  constexpr size_t INDENT = 4;
+  OS.indent(INDENT) << "ENTRY POINTS"
+                    << " " << Msg << " {\n";
+  for (const Function *F : C)
+    OS.indent(INDENT) << "  " << F->getName() << "\n";
+
+  OS.indent(INDENT) << "}\n";
+}
+
+void dumpEntryPoints(raw_ostream &OS, const Module &M,
+                     bool OnlyKernelsAreEntryPoints, std::string_view Msg) {
+  constexpr size_t INDENT = 4;
+  OS.indent(INDENT) << "ENTRY POINTS (Module)"
+                    << " " << Msg << " {\n";
+  for (const auto &F : M)
+    if (isEntryPoint(F, OnlyKernelsAreEntryPoints))
+      OS.indent(INDENT) << "  " << F.getName() << "\n";
+
+  OS.indent(INDENT) << "}\n";
+}
+
+void ModuleDesc::cleanup() {
+  // Externalize them so they are not dropped by GlobalDCE
```

Comment:
Externalize whom? I understand that it is a simplified version of what we currently have with for the sake of running `Internalize` pass here, but the comment seems to be off

Reply:
This is removed from the current patch.